### ISSUE TITLE
@Accessors does not generate @Override or @Deprecated when required

### ISSUE THREAD

**Issue by ivaniesta14** [eclipse/xtext-xtend#903]
Consider the following code:
```xtend
interface IFoo {
  abstract def Bar getBar()
}
class Baz implements IFoo {
  @Accessors Bar bar
}
```
In the compiled java class Baz, the method `getBar` is not tagged with `@Override`, triggering a warning.
I suspect this may also happen to the `@Data` annotation.

**Comment by cdietrich** [comment]
the processors currently dont check and add override annotation

**Comment by ivaniesta14** [comment]
Another related problem with the same cause: 
Marking the field as `@Deprecated` does not make the accessors deprecated, only the field.

**Comment by cdietrich** [comment]
@ivaniesta14 do you consider a special treatment for `@Deprecated` or do you have any criteria which annotations should be pulled over and which ones not

**Comment by ivaniesta14** [comment]
Ideally, the Java standard annotations (`@Deprecated`, `@Override` and `@SuppressWarnings`) and some nullness-related ones (notably `edu.umd.cs.findbugs.annotations.NonNull` and `Nullable` from Find/Spotbugs, `javax.annotation.Nonnull` and `CheckForNull` from JSR-305, and `org.eclipse.jdt.annotation.NonNull` and `Nullable` from Eclipse annotations) should be passed to the methods.
Alternatively, copying all annotations valid for both fields and methods is another option, but I think it may be inconvenient in some edge cases.

**Comment by szarnekow** [comment]
@ivaniesta14 I added an API that allows to ask for the overridden methods. Do you want to give it a try and file PR with the API incorporated into the active annotation compiler?

**Comment by ivaniesta14** [comment]
I would like to but I can't properly set up the project in Eclipse. I can commit the uncompiled (updated) Xtend code if needed.

**Comment by ivaniesta14** [comment]
Based on the `@Accessors` annotation and its processor I managed to create a temporary workaround:
```xtend
@Active(AccessorsOverrideProcessor)
@Target(ElementType.FIELD)
@Documented
annotation AccessorsOverride {
	AccessorType[] value = #[AccessorType.PUBLIC_GETTER, AccessorType.PUBLIC_SETTER]
	/**Generate {@code @Override} annotation on getter */
	boolean overrideGetter = false
	/**Generate {@code @Override} annotation on setter */
	boolean overrideSetter = false
	/**Annotations to delegate to the getter */
	Class<? extends Annotation>[] delegateGetter = #[]
	/**Annotations to delegate to the setter */
	Class<? extends Annotation>[] delegateSetter = #[]
	/**Annotations to delegate to the both the getter and the setter*/
	Class<? extends Annotation>[] delegate = #[]
	/**Use void setter instead of {@code this}-returning setter */
	boolean voidSetter = false
	/**Deprecate both setter and getter */
	boolean deprecate = false
}
class AccessorsOverrideProcessor extends AccessorsProcessor {
	override dispatch void transform(MutableFieldDeclaration it, extension TransformationContext context) {
		extension val util = new Util(context)
		val annot = accessorsOverrideAnnotation
		annot !== null && final
			? {
			annot.getBooleanValue('overrideSetter')
				? addWarning(
			'''
			Field «it» is final, but overrideSetter is true.
			Setting it has no effect, as no setter will be generated.''')
			annot.getBooleanValue('voidSetter')
				? addWarning(
			'''
			Field «it» is final, but voidSetter is true.
			Setting it has no effect, as no setter will be generated.''')
		}
		if(shouldAddGetter) addGetter(getGetterType.toVisibility)
		if(shouldAddSetter) addSetter(setterType.toVisibility)
	}

	static class Util extends AccessorsProcessor.Util {
		extension final TransformationContext cont

		new(TransformationContext context) {
			super(context)
			cont = context
		}

		def getAccessorsOverrideAnnotation(AnnotationTarget it) { findAnnotation(AccessorsOverride.findTypeGlobally) }
		def fieldOwner(MutableFieldDeclaration it) { if(static) declaringType.newTypeReference else "this" }
		def orObject(TypeReference it) { it !== null ? it : object }

		@SuppressWarnings('unchecked')
		override getGetterType(FieldDeclaration it) {
			val it = accessorsOverrideAnnotation
			it !== null
				? getEnumArrayValue("value").map[AccessorType.valueOf(simpleName)].findFirst[name.endsWith("GETTER")] ?:
				AccessorType.NONE
		}
		@SuppressWarnings('unchecked')
		override addGetter(MutableFieldDeclaration field, Visibility visibility) {
			field.validateGetter
			field.markAsRead
			field.declaringType.addMethod(field.getterName) [
				primarySourceElement = field.primarySourceElement
				addAnnotation(newAnnotationReference(Pure))
				val annot = field.accessorsOverrideAnnotation
				annot !== null
					? {
					// Add eplicit Override
					annot.getBooleanValue('overrideGetter') ? addAnnotation(newAnnotationReference(Override))
					// Add explicit Deprecated
					annot.getBooleanValue('deprecate') ? addAnnotation(newAnnotationReference(Deprecated))
					// Add delegated annotations
					(annot.getClassArrayValue('delegateGetter') + annot.getClassArrayValue('delegate')).forEach [ an |
						addAnnotation(newAnnotationReference(an.name))
					]
				}
				// Deprecate Accessors if field is deprecated
				field.annotations.map[annotationTypeDeclaration].exists[qualifiedName == Deprecated.name]
					? addAnnotation(newAnnotationReference(Deprecated))
				returnType = field.type.orObject
				body = '''return «field.fieldOwner».«field.simpleName»;'''
				static = field.static
				it.visibility = visibility
			]
		}

		@SuppressWarnings('unchecked')
		override getSetterType(FieldDeclaration it) {
			val it = accessorsOverrideAnnotation
			it !== null
				? getEnumArrayValue("value").map[AccessorType.valueOf(simpleName)].findFirst[name.endsWith("SETTER")] ?:
				AccessorType.NONE
		}
		@SuppressWarnings('unchecked')
		override addSetter(MutableFieldDeclaration field, Visibility visibility) {
			field.validateSetter
			field.declaringType.addMethod(field.setterName) [
				primarySourceElement = field.primarySourceElement
				val annot = field.accessorsOverrideAnnotation
				annot !== null
					? {
					// Add explicit override
					annot.getBooleanValue('overrideSetter') ? addAnnotation(newAnnotationReference(Override))
					// Add explicit deprecated
					annot.getBooleanValue('deprecate') ? addAnnotation(newAnnotationReference(Deprecated))
					// Add delegated annotations
					(annot.getClassArrayValue('delegateSetter') + annot.getClassArrayValue('delegate')).forEach [ an |
						addAnnotation(newAnnotationReference(an.name))
					]
					// Use void setter?
					field.static || annot.getBooleanValue('voidSetter')
						? {
						returnType = primitiveVoid;
						val param = addParameter(field.simpleName, field.type.orObject)
						body = '''«field.fieldOwner».«field.simpleName» = «param.simpleName»;'''
					}
						: {
						returnType = field.declaringType.newTypeReference
						val param = addParameter(field.simpleName, field.type.orObject)
						body = '''
							«field.fieldOwner».«field.simpleName» = «param.simpleName»;
							return this;
						'''
					}
				}
					: { // Ensure void setter
					returnType = primitiveVoid;
					val param = addParameter(field.simpleName, field.type.orObject)
					body = '''«field.fieldOwner».«field.simpleName» = «param.simpleName»;'''
				}
				// Deprecate Accessors if field is deprecated
				field.annotations.map[annotationTypeDeclaration].exists[qualifiedName == Deprecated.name]
					? addAnnotation(newAnnotationReference(Deprecated))
				static = field.static
				it.visibility = visibility
			] // field.declaringType.addMethod(...)[...]
		} // override addSetter
	} // static class Util
} //class AccessorsOverrideProcessor
```
Currently does not automatically add the `@Override` annotation but does add `@Deprecated` and has the posibility to delegate annotations to the getter and setter.
Also has an option to toggle between void setter and `this`-returning setter.

I will try to add it to the repository when I manage to fix the problems I previously had setting it up locally.

**Comment by miklossy** [comment]
Hi @ivaniesta14 ! Do you have any progress on this topic?

**Comment by ivaniesta14** [comment]
Yes! The code is almost ready, but I have been incredibly busy recently. I'll try to push ASAP.

**Comment by miklossy** [comment]
> Yes! The code is almost ready, but I have been incredibly busy recently. I'll try to push ASAP.

Nice to hear that!

**Comment by cdietrich** [comment]
@szarnekow @ivaniesta14 anything left?

**Comment by cdietrich** [comment]
fixed in 2.23
